test/obfuscator-spec.ts

```diff
@@ -0,0 +1,208 @@
+  fields: {
+    _id: { is: 'mongoid' },
+    firstName: { is: 'string' },
+    lastName: { is: 'string' },
+    ip_address: { is: 'string' }
+  }
+});
+
+const logObj = (input: any) => {
+  console.log(JSON.stringify(input));
+};
+
+test.before('Setup Database', t => {
+  dbManager = new DBManager();
+  return dbManager.start();
+});
+
+/**
+ * TODO: Figure out way to set up context for multiple tests
+ * Determine that environment is correctly setup and Tyranid correctly configured
+ */
+test.serial('Collection sould have all user entries', async t => {
+  const user = await User.findAll({ query: {}, count: true});
+  const isValid = user.count === TestDataSet.length;
+
+  t.true(isValid, 'Could not access test data');
+});
+
+test.serial('Should be able to retrieve document by explicit collection or byName', async t => {
+  const user = await User.findOne({ query: {} });
+  const user2 = await Tyr.byName.user.findOne({ query: {} });
+
+  const isValid = (JSON.stringify(user) === JSON.stringify(user2));
+
+  t.true(isValid, 'Could not retrieve same objects by Collection.findOne & Tyr.byName.Collection.findOne');
+});
+
+test.serial('Field Definition should save encryptable flag', async t => {
+  const user = await User.findOne({ query: {} });
+  const userFields = user.$model.def.fields;
+  const isObfuscateable = userFields.lastName.def.obfuscateable;
+  const isNotObfuscateable = userFields.createdAt.def.hasOwnProperty('obfuscateable');
+
+  t.true(isObfuscateable, 'Does not have obfuscateable flag, extending FieldInstance definition failed. ');
+  t.false(isNotObfuscateable, 'Non-obfuscateable field has flag set to true');
+});
+
+test('Should have obfuscate function on Tyr namespace', t => {
+  const valid = Tyr.obfuscate && typeof Tyr.obfuscate === 'function';
+  t.true(valid, 'Tyranid namespace does not have obfuscate function');
+});
+
+/**
+ * Test Obfuscation
+ * 
+ * 
+ * TODO Test Cases
+ *   1. Should complete exporting, encryption, and recovery with datasize that requires more than one batch
+ *   2. ALL error scenarios and recovery
+ *   3. Complex datatypes, arrays, objects, etc
+ *   4. Test null data types, missing fields
+ */
+
+test.serial('Should copy obfuscateable data to temp collection', async t => {
+  const collection = await createFreshCopyOfData(); 
+  const copiedData = (await collection.findAll({ query: {  } }));
+  t.deepEqual(JSON.stringify(copiedData), ExpectedResults.CopiedObfuscateableData, 'Obfuscateable data incorrectly copied to new collection');
+});
+
+test.serial('Should encrypt given collection', async t => {
+  //Create collection of values
+  const collection = await createFreshCopyOfData();
+
+  const resultTBD = await Tyr.encryptCollection(collection, masterKey);
+
+  const records = await collection.findAll({ query: {} });
+  t.true(records.length === 10, 'All records still in collection');
+  t.notDeepEqual(JSON.stringify(records), ExpectedResults.CopiedObfuscateableData, 'Data not encrypted');
```

Comment:
Future: This assert seems a little prone to false positives. For example, if anyone accidentally adds to the users dataset but not the expected results, the test would always pass. Compounded slightly with the stringification: if stringify ever returns a format slightly different from your string literal expected result, same issue. What was the purpose of using stringify by the way? Were you running into issues with a regular deep comparison?

Reply:
Very true, I was trying to quickly replicate how DBunit asserts database result batches. Also I wasn't sure how to account for the unique values generated in a simplistic comparison.